lib/elixir/lib/json.ex

```diff
@@ -0,0 +1,390 @@
+
+    try do
+      :elixir_json.decode(binary, acc, Map.new(decoders))
+    catch
+      :error, :unexpected_end ->
+        {:error, {:unexpected_end, byte_size(binary)}}
+
+      :error, {:invalid_byte, byte} ->
+        {:error, {:invalid_byte, offset(__STACKTRACE__), byte}}
+
+      :error, {:unexpected_sequence, bytes} ->
+        {:error, {:unexpected_sequence, offset(__STACKTRACE__), bytes}}
+    end
+  end
+
+  defp offset(stacktrace) do
+    with [{_, _, _, opts} | _] <- stacktrace,
+         %{cause: %{position: position}} <- opts[:error_info] do
+      position
+    else
+      _ -> 0
+    end
+  end
+
+  @doc ~S"""
+  Decodes the given JSON but raises an exception in case of errors.
+
+  Returns the decoded content. See `decode/1` for possible errors.
+
+  ## Examples
+
+      iex> JSON.decode!("[null,123,\"string\",{\"key\":\"value\"}]")
+      [nil, 123, "string", %{"key" => "value"}]
+  """
+  @spec decode!(binary()) :: term()
+  def decode!(binary) when is_binary(binary) do
+    case decode(binary) do
+      {:ok, decoded} ->
+        decoded
+
+      {:error, {:unexpected_end, offset}} ->
+        raise JSON.DecodeError,
+          message: "unexpected end of JSON binary at position (byte offset) #{offset}",
+          data: binary,
+          offset: offset
+
+      {:error, {:invalid_byte, offset, byte}} ->
+        raise JSON.DecodeError,
+          message: "invalid byte #{byte} at position (byte offset) #{offset}",
+          data: binary,
+          offset: offset
+
+      {:error, {:unexpected_sequence, offset, bytes}} ->
+        raise JSON.DecodeError,
+          message: "unexpected sequence #{inspect(bytes)} at position (byte offset) #{offset}",
+          data: binary,
+          offset: offset
+    end
+  end
+
+  @doc ~S"""
+  Encodes the given term to JSON as a binary.
+
+  The second argument is a function that is recursively
+  invoked to encode a term.
+
+  ## Examples
+
+      iex> JSON.encode!([123, "string", %{key: "value"}])
+      "[123,\"string\",{\"key\":\"value\"}]"
+
+  """
+  @spec encode!(a, (a -> iodata())) :: binary() when a: var
+  def encode!(term, encoder \\ &encode_value/2) do
+    IO.iodata_to_binary(encoder.(term, encoder))
+  end
+
+  @doc ~S"""
+  Encodes the given term to JSON as an iodata.
+
+  This is the most efficient format if the JSON is going to be
+  used for IO purposes.
+
+  The second argument is a function that is recursively
+  invoked to encode a term.
+
+  ## Examples
+
+      iex> data = JSON.encode_to_iodata!([123, "string", %{key: "value"}])
+      iex> IO.iodata_to_binary(data)
+      "[123,\"string\",{\"key\":\"value\"}]"
+
+  """
+  @spec encode!(a, (a -> iodata())) :: iodata() when a: var
+  def encode_to_iodata!(term, encoder \\ &encode_value/2) do
+    encoder.(term, encoder)
+  end
+
+  @doc """
+  This is the default function used to recursively encode each value.
+  """
+  def encode_value(value, encoder) when is_atom(value) do
```

Comment:
It is the same as the protocol, inlined for performance. But I am not a fan of its name. Suggestions? What about `encode_callback`? or `default_encode`? or `default_encode_callback`?

Reply:
Given it isn't technically an `@callback`, even if it is a generic term `_callback` might be confusing since its specific meaning in OTP.
`default_encode` captures the intent well I think.